starlette/applications.py

```diff
@@ -109,7 +111,9 @@ def url_path_for(self, name: str, **path_params: str) -> URLPath:
 
     async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
         scope["app"] = self
-        await self.middleware_stack(scope, receive, send)
+        task_group = scope["task_group"] = anyio.create_task_group()
+        async with task_group:
+            await self.middleware_stack(scope, receive, send)
```

Comment:
Something we need to be aware of here, that's *kinda* silly, but also is at least worth pointing out...

This is on the critical path, so it'll have some impact on the benchmarking.

Now, I'm acutely aware that the micro-benchmarking is a big ol' pile of nonsense. But also it does, tediously, lead some folks into particular evaluations.

When I started Starlette one of the design arguments that I felt needed to be knocked down was that integrated server+application frameworks would be necessarily "faster" than frameworks with a clean server / application interface split. In this case ASGI.

To that extent it was important that Starlette ought to be able to equal sanic or aiohttp when evaluated with microbenchmarks, however tedious that might be.

Adoptiong this would likely change that.

Which might well be *okay*, but it's something we need to consider.

Reply:
I can provide some benchmarks for this conversation.  Are there any that you recommend as meaningful?

Reply:
@tomchristie I created a branch here: https://github.com/uSpike/FrameworkBenchmarks/tree/starlette-anyio

See results: https://www.techempower.com/benchmarks/#section=test&shareid=c92a1338-0058-486c-9e47-c92ec4710b6a&hw=ph&test=query&a=2

There is a performance penalty to using anyio.  The weighted composite score is 173 vs 165, so a 4.5% drop in performance.

I'll add this result to the top comment.

Reply:
Interestingly, the multiple queries test was ~25% faster with anyio.

Reply:
I ran the benchmark again against sanic and aiohttp, see https://www.techempower.com/benchmarks/#section=test&shareid=a2023abc-ee86-458f-a0e9-849b5c9f282f&hw=ph&test=composite&a=2

Reply:
I only see starlette and starlette-anyio. How do I run these benchmarks locally?

Reply:
@agronholm if you click the other test tabs they should show up.

You can run locally with
```
./tfb --test starlette starlette-anyio sanic aiohttp
```

Reply:
It's hard to get reproducible results on my local machine.  I ran it again now with @graingert changes to make task groups per-request, and now starlette-anyio is much closer to starlette performance: https://www.techempower.com/benchmarks/#section=test&shareid=d1edb7b5-09f0-43c0-a4cb-cca51f87801c

Reply:
task groups were per request before, they were just attached to the app